### ISSUE TITLE
Docker: Upgrade RawTherapee to fix ProRAW support

### ISSUE THREAD

**Issue by charles-997** [photoprism/photoprism#2291]
**What does not work as described in the documentation?**
Images created from my various iPhone models that support the Apple ProRAW image type are not rendered properly in the PhotoPrism WebUI. See example:
<img width="1680" alt="Screen Shot 2022-04-29 at 2 49 30 PM" src="https://user-images.githubusercontent.com/52077853/166007867-9d23a355-0208-4d7c-860d-f8b1d70af108.png">

**How can we reproduce it?**

1. Import images taken on an iPhone using Apple's [ProRAW image format](https://support.apple.com/en-us/HT211965) to your PP library.
2. Browse the images via http://photoprism.<domain>.net/browse
3. Click on any of the iPhone ProRAW image to view full screen. (Note the image thumbnails also render improperly.)

_If reporting an Import, Indexing, or Performance issue, please include the number and type of pictures in your library,
as well as any configuration options you have changed e.g. for thumbnail quality._  There are about 60k images in my library.

**What behavior do you expect?**

I expect the image to render in a viewable state similar to how Sony .ARW RAW files or Canon .CR2 RAW files are rendered.

**What could be the cause of your problem?**

As I understand, PP uses upstream software - DarkTable & RawTherapee to render JPEG previews from RAW images.  Darktable appears to not yet support iPhone ProRAW files - refer to [DarkTable GitHub Issue](https://github.com/darktable-org/darktable/issues/7578)
Is there an alternative way PhotoPrism could render these images?  I do not know the number of PhotoPrism users who 1. use an iPhone for photography and 2. use the Apple ProRAW file type, however I would be surprised if it were not a significant percentage of users.

Try to determine the cause of your problem before submitting a bug report:
https://docs.photoprism.app/getting-started/troubleshooting/

**Can you provide us with example files for testing, error logs, or screenshots?**

Please add example files or screenshots that help to reproduce your problem.
You may also send files via email or share a download link:
https://photoprism.app/contact

See example in PhotoPrism
<img width="1680" alt="Screen Shot 2022-04-29 at 2 49 30 PM" src="https://user-images.githubusercontent.com/52077853/166007867-9d23a355-0208-4d7c-860d-f8b1d70af108.png">

Learn how to diagnose frontend issues:
https://docs.photoprism.app/getting-started/troubleshooting/browsers/

NOTE:
- You have to provide sample files and logs if it is an IMPORT, INDEXING, or METADATA issue, otherwise we will not be able to process your report
- If it is an IMPORT issue, please also provide an archive with affected files before importing them so that it's possible to reproduce your issue

**Which software versions do you use?**

- PhotoPrism Architecture & Build Number (220302-0059f429-Linux-AMD64)
- Database Type & Version (MariaDB)
- Operating System Types & Versions (PhotoPrism running on Linux in Docker)
- Browser Types & Versions (Firefox 99.0.1, macOS 12.3.1)
- Browser Plugins & Ad Blockers (Bitwarden)

The app version / build number can be found in *Settings* when you scroll to the bottom.
MySQL 8 is not officially supported anymore, see System Requirements.

NOTE:
- Always provide database and operating system details if it is a Backend, Import, or Indexing issue
- Always provide web browser and operating system details if it is a Frontend issue
- If it is a Performance problem, you must provide ALL information

**On what kind of device is PhotoPrism installed?**

This is especially important if you are reporting a Performance, Import, or Indexing problem. You can skip this if you are reporting a problem found on our public demo, or if it is a completely unrelated issue, such as a broken page layout.

- Device / Processor Type (Unraid custom server build, i5-8400)
- Physical Memory & Swap Space (48GB physical memory, unclear swap size)
- Storage Type (TrueNAS ZFS Array mounted to linux server via NFS.  Linux server provides the storage to Docker where PP is running) 
- anything else that might be helpful

NOTE:
- Always provide device, memory, and storage details when you have a Backend, Performance, Import, or Indexing issue

**Do you use a Reverse Proxy, Firewall, VPN, or CDN?**

Firewall- pfSense 2.5.2 on dedicated physical host.  Runs a few remote access VPNs. These should be out of scope of my photoprism issue.

NOTE:
- Always provide this information when you have a Reliability, Performance, or Frontend problem, such as failed uploads, connection errors, broken thumbnails, or video playback issues
- If you are using NGINX, also provide its configuration and/or consider asking the NGINX community for advice as we do not specialize in supporting their product: https://docs.photoprism.app/getting-started/proxies/nginx/

**Comment by lastzero** [comment]
Known issue, not properly supported by Darktable and Rawtherapee yet. Should work with Sips on macOS without Docker when you build PhotoPrism from source.

**Comment by giacy86** [comment]
Hi. Latest [RawTherapee dev build correctly handles Apple ProRaw](https://github.com/Beep6581/RawTherapee/pull/6214) iPhone dng images (though the thumbnails are a little washed out but still better than nothing). I have tested with the [dev AppImage build](https://github.com/Beep6581/RawTherapee/releases/download/nightly-github-actions/RawTherapee_dev_release.AppImage) and worked fine. 
So i have copied the rawtherapee-cli bin and all related libraries to the docker container with `docker cp`. I have also copied a dng image for testing inside the container, entered with `photoprism bash` and tried to convert the dng with `rawtherapee-cli -c image.dng` (launching `rawtherapee-cli` inside the container show the correct dev build). Then i have copied the converted jpg to my home and the image looks fine. 
However, when i launch a `index -f` or `thumbs -f` the thumbnails are corrupted. Darktable is disabled in settings.
Can you point me in the right direction?

**Comment by lastzero** [comment]
This request came a bit too late, as we have just released. For reference, this is the version that ships with our default Docker image:

https://packages.debian.org/bookworm/rawtherapee

**Comment by Beurling** [comment]
The problem is diffrent.Mine dng photos are all missing in the web view

**Comment by lastzero** [comment]
It seems for you the conversion is not working at all, which ultimately must be fixed in the same way.

**Comment by Beurling** [comment]
Thank you.

**Comment by lastzero** [comment]
While RawTherapee has not been updated yet, PhotoPrism can now extract JPEGs from ProRAW files so they can be indexed properly - at least in my tests.

**Comment by ark-** [comment]
Nice work on the latest Apple support. Thank you team. 

I have a bunch of thumbnails like the OP also. If I reindex those using the latest development build will my thumbs be fixed?

**Comment by nickgomez19** [comment]
@lastzero , how do you use this new capability in photoprism?

**Comment by lastzero** [comment]
We will provide a development preview and then a new stable version as soon as I get to it!

**Comment by lastzero** [comment]
https://demo.photoprism.app/ has been updated with the latest changes! Feel free to test your HEIC and DNG files via web upload on the demo while we get the Development Preview ready for testing (it still has some migration issues so you wouldn't be able to log in).

**Comment by epoblaguev** [comment]
> https://demo.photoprism.app/ has been updated with the latest changes! Feel free to test your HEIC and DNG files via web upload on the demo while we get the Development Preview ready for testing (it still has some migration issues so you wouldn't be able to log in).

I uploaded a 48MP proraw file, and it seems to have been processed correctly.

**Comment by lastzero** [comment]
In fact, we don't process it, but extract the embedded JPEG from the DNG files. They seem to have the full resolution, or at least a very high resolution, and it's faster too.

**Comment by Beurling** [comment]
> https://demo.photoprism.app/ has been updated with the latest changes! Feel free to test your HEIC and DNG files via web upload on the demo while we get the Development Preview ready for testing (it still has some migration issues so you wouldn't be able to log in).

Thank you very much!
Interesting things is my dng photo works well in the demo.app.But still show black in my docker.I notice that the version of the demo app is 
[PhotoPrism® Plus](https://demo.photoprism.app/about)
 Build [221014-3bad6820d](https://docs.photoprism.app/release-notes/) 
but mine is 
[PhotoPrism® CE](http://10.0.2.3:2342/about)
Build [220901-f493607b0](https://docs.photoprism.app/release-notes/)

I wonder should I wait for the update in the future or become a plus user so I  can have this dng funtion works?

Thanks again!

**Comment by graciousgrey** [comment]
@Beurling DNG support will also be available in the Community Edition. Of course we would still be happy to have you as a [sponsor](https://photoprism.app/membership)  :)

**Comment by Beurling** [comment]
> @Beurling DNG support will also be available in the Community Edition. Of course we would still be happy to have you as a [sponsor](https://photoprism.app/membership) :)

Thanks.I will consider about it.

**Comment by danielunderwood** [comment]
It may be worth noting that upgrading the version fixed the issue with new uploads, but I had to do something like the following to fix previously uploaded photos:
```bash
find /path/to/photoprism/sidecar -type f -name "*.dng.*" -exec rm {} \;
photoprism convert -f affected/photos
photoprism index -f affected/photos
photoprism thumbs -f affected/photos
```

I'm not sure if it was one particular step here that fixed the issue, but I ended up trying all of them before the issue was resolved.

**Comment by graciousgrey** [comment]
_"Indexing is also necessary to find and view HEIC, DNG, and AVIF images that were previously unsupported or had errors. In some cases with incorrectly converted images, it may be necessary to recreate the JPEG sidecar files by running the photoprism convert -f command [in a terminal](https://docs.photoprism.app/getting-started/docker-compose/#command-line-interface) before starting the rescan. To regenerate your thumbnails, run photoprism thumbs -f."_

--> https://docs.photoprism.app/release-notes/#november-2-2022

**Comment by ErikGab** [comment]
Related but not identical issue. A large portion of my raw images from a Canon EOS 70D renders only black and bright green. This is not an issue in photoprism, it was a bug in rawtherapee introduced in 5.8 and fixed in 5.9.

I have tried to add the rawtherapee 5.9 appimage and the extracted version of the appimage to the photoprism docker container using a volume in my docker-compose.yml in combination with setting the PHOTOPRISM_RAWTHERAPEE_BIN environment variable value to a string matching the path and filename that I volumed in. But no success so far.

Is there a way of getting photoprism to use rawtherapee 5.9?

**Comment by lastzero** [comment]
We currently use the version that comes with Ubuntu. But we can also install from a another source like with Darktable.

**Comment by lastzero** [comment]
@ErikGab AFAIK Darktable should be used by default for Canon RAW files. Is it disabled?

Unfortunately, I couldn't find any RawTherapee 5.9 packages for the version of Ubuntu we are currently using (22.04 LTS):

- https://software.opensuse.org/package/rawtherapee

However, the upcoming 23.04 release should include RawTherapee 5.9, so we could wait a bit and then upgrade:

- https://launchpad.net/ubuntu/lunar/+source/rawtherapee

Using the version included in the distribution also ensures that the package is available for all architectures, not just AMD64/x86.

**Comment by mmaravillo** [comment]
**EDIT:  Added apt pinning to jammy.**

You can also add Ubuntu's Lunar Lobster repository in the container so that you can upgrade rawtherapee.

~~Don't run **apt upgrade**, or else it will upgrade the whole OS of the container.~~
```
mike@WALNUT:~/build/photoprism$ make root-terminal
docker compose exec -u root photoprism bash
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ dpkg -l | grep rawtherapee
ii  rawtherapee                          5.8-3                                   amd64        raw image converter and digital photo processor
ii  rawtherapee-data                     5.8-3                                   all          raw image converter and digital photo processor (data files)
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ echo deb http://archive.ubuntu.com/ubuntu/ lunar main >> /etc/apt/sources.list.d/rawtherapee.list
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ echo deb http://archive.ubuntu.com/ubuntu/ lunar universe >> /etc/apt/sources.list.d/rawtherapee.list
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ echo 'APT::Default-Release "jammy";' > /etc/apt/apt.conf.d/80release
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ cat << EOF > /etc/apt/preferences.d/rawtherapee.pref
Package: rawtherapee rawtherapee-data libatkmm-1.6-1v5 libcairomm-1.0-1v5 libglibmm-2.4-1v5 libgtkmm-3.0-1v5 libpangomm-1.4-1v5 liblensfun1 liblensfun-data-v1
Pin: release n=lunar
Pin-Priority: 990
EOF
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ apt update
...
Fetched 45.8 MB in 23s (2029 kB/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
535 packages can be upgraded. Run 'apt list --upgradable' to see them.
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ apt install rawtherapee
...
The following additional packages will be installed:
  libatkmm-1.6-1v5 libcairomm-1.0-1v5 libglibmm-2.4-1v5 libgtkmm-3.0-1v5 liblensfun-data-v1 liblensfun1 libpangomm-1.4-1v5 rawtherapee-data
The following packages will be upgraded:
  libatkmm-1.6-1v5 libcairomm-1.0-1v5 libglibmm-2.4-1v5 libgtkmm-3.0-1v5 liblensfun-data-v1 liblensfun1 libpangomm-1.4-1v5 rawtherapee rawtherapee-data
9 upgraded, 0 newly installed, 0 to remove and 526 not upgraded.
Need to get 23.5 MB of archives.
After this operation, 14.4 MB of additional disk space will be used.
...
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ dpkg -l | grep rawtherapee
ii  rawtherapee                          5.9-1                                   amd64        raw image converter and digital photo processor
ii  rawtherapee-data                     5.9-1                                   all          raw image converter and digital photo processor (data files)
root@230111-jammy:/go/src/github.com/photoprism/photoprism$ rawtherapee -v
RawTherapee, version 5.9
```

**Comment by ErikGab** [comment]
Hi, Thanks for the replies @lastzero @mmaravillo . I understand that it won't have much of a priority for the Photoprism team since the actual issue is in another software, and it will be fixed by just waiting.

Anyway, thanks for a great job with Photoprism. Was initially trying to save some money moving away from subscriptions, but after using RawTherapee in combination with Photoprism I prefer the workflow and features more than the Adobe way I have been using. To bad I run into the Rawtherapee bug when moving back in my library timeline to a previous camera I once had.

I will try to get it working by adjusting the container. @mmaravillo thanks for the tip using lunar.

**Comment by lastzero** [comment]
Actually, file type support has a high priority for us: https://photoprism.app/kb/oss

**Comment by ErikGab** [comment]
@lastzero @mmaravillo just to let you know. Installed 5.9 in the container as @mmaravillo described, worked like a charm.
New imported images are developed as they should. What is the recommended way to force recreation of jpgs for the raw images that has already been converted prior the 5.9 installation?

**Comment by lastzero** [comment]
You can use the convert command in a terminal for this or delete the sidecar files and start a complete rescan. The development preview additionally supports specifying file extensions, which would be the fastest option.

**Comment by mmaravillo** [comment]
> @lastzero @mmaravillo just to let you know. Installed 5.9 in the container as @mmaravillo described, worked like a charm. New imported images are developed as they should.

Btw, @ErikGab if you check PR #3102 there is code there to pin the OS to jammy (Ubuntu 22.04).  This way, if you accidentally ran apt upgrade or dist-upgrade, the container will not be upgraded to lunar.

```
echo 'APT::Default-Release "jammy";' > /etc/apt/apt.conf.d/80release
cat << EOF > /etc/apt/preferences.d/rawtherapee.pref
Package: rawtherapee rawtherapee-data libatkmm-1.6-1v5 libcairomm-1.0-1v5 libglibmm-2.4-1v5 libgtkmm-3.0-1v5 libpangomm-1.4-1v5 liblensfun1 liblensfun-data-v1
Pin: release n=lunar
Pin-Priority: 990
EOF
```

**Comment by lastzero** [comment]
Our new preview build is based on Ubuntu 23.04, which includes RawTherapee 5.9, Darktable 4.2 and FFmpeg 5.1. We appreciate any help with testing! You can also use our demo at https://demo.photoprism.app for this.

**Comment by lastzero** [comment]
Full Release Notes:
- https://docs.photoprism.app/release-notes/#development-preview
